sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/LabeledMetricNameUtils.java

```diff
@@ -0,0 +1,122 @@
+   * formatted as:
+   *
+   * <p>'{baseName}-{metricLabelKey1}:{metricLabelVal1};...{metricLabelKeyN}:{metricLabelValN};'
+   */
+  public static class MetricNameBuilder {
+    private final StringBuilder labeledNameBuilder;
+
+    private MetricNameBuilder(String baseName) {
+      this.labeledNameBuilder = new StringBuilder(baseName + METRIC_NAME_DELIMITER);
+    }
+
+    public static MetricNameBuilder baseNameBuilder(String baseName) {
+      return new MetricNameBuilder(baseName);
+    }
+
+    /**
+     * Add a metric label KV pair to the metric name. Calling {@code addLabel} multiple times with
+     * the same {@code key} will lead to unexpected results when parsing metric names.
+     */
+    public void addLabel(String key, String value) {
+      this.labeledNameBuilder
+          .append(key)
+          .append(METRIC_KV_DELIMITER)
+          .append(value)
+          .append(LABEL_DELIMITER);
+    }
+
+    public MetricName build(String metricNamespace) {
+      return MetricName.named(metricNamespace, labeledNameBuilder.toString());
+    }
+  }
+
+  @AutoValue
+  public abstract static class ParsedMetricName {
+    public abstract String getBaseName();
+
+    public abstract Map<String, String> getMetricLabels();
+
+    public static ParsedMetricName create(String baseName, Map<String, String> metricLabels) {
+      return new AutoValue_LabeledMetricNameUtils_ParsedMetricName(baseName, metricLabels);
+    }
+
+    public static ParsedMetricName create(String baseName) {
+      ImmutableMap<String, String> emptyMap = ImmutableMap.of();
+      return new AutoValue_LabeledMetricNameUtils_ParsedMetricName(baseName, emptyMap);
+    }
+  }
+
+  /**
+   * Parse a 'metric name' String that was created with 'MetricNameBuilder'. The input string should
+   * be formatted as.
+   *
+   * <p>'{baseName}*{metricLabelKey1}:{metricLabelVal1};...{metricLabelKeyN}:{metricLabelValN};'
+   *
+   * @param metricName
+   * @return Returns a ParsedMetricName object if the input string is properly formatted. If the
+   *     input string is empty or malformed, returns an empty value.
+   */
+  public static Optional<ParsedMetricName> parseMetricName(String metricName) {
+    if (metricName.isEmpty()) {
+      return Optional.empty();
+    }
+
+    List<String> metricNameSplit =
+        Splitter.on(METRIC_NAME_DELIMITER).limit(2).splitToList(metricName);
+
+    if (metricNameSplit.size() == 0) {
+      return Optional.empty();
+    }
+
+    if (metricNameSplit.size() == 1) {
+      return Optional.of(ParsedMetricName.create(metricNameSplit.get(0)));
+    }
+
+    Splitter.MapSplitter splitter =
+        Splitter.on(LABEL_DELIMITER).omitEmptyStrings().withKeyValueSeparator(METRIC_KV_DELIMITER);
+    try {
+      Map<String, String> labels = splitter.split(metricNameSplit.get(1));
+      return Optional.of(ParsedMetricName.create(metricNameSplit.get(0), labels));
+    } catch (IllegalArgumentException e) {
```

Comment:
should we log something here?

Reply:
or maybe just move the return statement below into here to show that we only do the latter when an an exception is thrown